lib/thread.c

```diff
@@ -1355,30 +1358,42 @@ struct thread *thread_fetch(struct thread_master *m, struct thread *fetch)
 		memcpy(m->handler.copy, m->handler.pfds,
 		       m->handler.copycount * sizeof(struct pollfd));
 
-		pthread_mutex_unlock(&m->mtx);
-		{
-			num = fd_poll(m, m->handler.copy, m->handler.pfdsize,
-				      m->handler.copycount, tw);
-		}
-		pthread_mutex_lock(&m->mtx);
+		/*
+		 * Attempt to flush ready queue before going into poll().
+		 * This is performance-critical. Think twice before modifying.
+		 */
+		if (m->ready.count == 0 || m->tick_since_io >= MAX_TICK_IO) {
+			pthread_mutex_unlock(&m->mtx);
+			{
+				m->tick_since_io = 0;
+				num = fd_poll(m, m->handler.copy,
+					      m->handler.pfdsize,
+					      m->handler.copycount, tw);
+			}
+			pthread_mutex_lock(&m->mtx);
+
+			/* Handle any errors received in poll() */
+			if (num < 0) {
+				if (errno == EINTR) {
+					pthread_mutex_unlock(&m->mtx);
+					/* loop around to signal handler */
+					continue;
+				}
 
-		/* Handle any errors received in poll() */
-		if (num < 0) {
-			if (errno == EINTR) {
+				/* else die */
+				zlog_warn("poll() error: %s",
+					  safe_strerror(errno));
 				pthread_mutex_unlock(&m->mtx);
-				continue; /* loop around to signal handler */
+				fetch = NULL;
+				break;
 			}
 
-			/* else die */
-			zlog_warn("poll() error: %s", safe_strerror(errno));
-			pthread_mutex_unlock(&m->mtx);
-			fetch = NULL;
-			break;
-		}
+			/* Since we could have received more cancellation
+			 * requests during poll(), process those */
+			do_thread_cancel(m);
 
-		/* Since we could have received more cancellation requests
-		 * during poll(), process those */
-		do_thread_cancel(m);
+		} else
+			m->tick_since_io++;
```

Comment:
I actually don't mind, but it seems this is [not following linux kernel code style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#placing-braces-and-spaces). I'm curious: was this output from clang script?